### PR TITLE
Confirm de-dupe of license encumbrance privilege update records #1050

### LINKED ISSUE #1050: double check adverse actionID when creating encumbrance_lifted and encumbrance updates to confirm that they it is not a duplicate.
### Tasks
- [ ] TODO
- [ ] Automated tests
- [ ] API Docs
- [ ] Postman collection
- [ ] PR opened with labels / reviewers / assignee / linked-issue

### PR DESCRIPTION
**Pull request by ChiefStief**
### Requirements List
-

### Description List
- Prevented creation of duplicate encumbrance update records by adding pre-checks that skip creating redundant update entries during home state license encumbrance => associated privilege flow
- Updated tests

### Testing List
- For API configuration changes: CDK tests added/updated in `backend/compact-connect/tests/unit/test_api.py`
- Code review

Closes #1050 

## Summary by CodeRabbit

* **Bug Fixes**
  * Added deduplication/pre-checks in encumbrance and lifting flows to prevent duplicate privilege-update records on retries and ensure idempotent processing.

* **Tests**
  * Added comprehensive tests validating retry/idempotency, preservation of original update records, single-time lifting updates, notification edge cases, and multiple/missing-privilege scenarios.

* **Refactor**
  * Standardized adverse-action identifier types to UUID across affected interfaces and docs.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: backend/compact-connect/lambdas/python/common/cc_common/data_model/data_client.py [hunk 1]
```diff
@@ -2733,6 +2754,27 @@ def encumber_home_jurisdiction_license_privileges(
...
         for encumbered_privilege in previously_encumbered_privileges_associated_with_license:
+            # Check if an update record already exists for this adverse action
+            # to avoid creating duplicate update records if the event flow is re-run
+            existing_updates = provider_user_records.get_update_records_for_privilege(
+                jurisdiction=encumbered_privilege.jurisdiction,
+                license_type=encumbered_privilege.licenseType,
+                filter_condition=lambda update: (
+                    update.updateType == UpdateCategory.ENCUMBRANCE
+                    and update.encumbranceDetails is not None
+                    and str(update.encumbranceDetails.get('adverseActionId')) == adverse_action_id
```
Comment (jusdino) [inline backend/compact-connect/lambdas/python/common/cc_common/data_model/data_client.py]: We've been working on tidying up our object type handling throughout the codebase. When we have a UUID, we should deserialize it early and handle it as a `UUID`, rather than a `str`, only serializing again to a `str` as late as possible before it leaves the system. Similarly, our date/datetime objects should be processed as the more intelligent types, not strings. To that end, rather than cast `adverseActionId` here to a `str,` can you instead chase this `adverse_action_id` up the call chain and convert it to a UUID where it comes in?

File: backend/compact-connect/lambdas/python/data-events/tests/function/test_encumbrance_events.py [hunk 2]
```diff
@@ -3031,3 +3031,250 @@ def test_license_encumbrance_lifting_notification_listener_skips_notifications_w
...
+
+        message = self._generate_license_encumbrance_message()
+        event = self._create_sqs_event(message)
+
+        # Execute the handler FIRST TIME
+        license_encumbrance_listener(event, self.mock_context)
+
+        # Verify privilege update record was created
+        serialized_privilege = privilege.serialize_to_database_record()
+        privilege_update_records = self._provider_table.query(
```
Comment (jusdino) [inline backend/compact-connect/lambdas/python/data-events/tests/function/test_encumbrance_events.py]: We're also trying to move away from 'peeking under the hood' in tests, where possible (as in querying directly with hard-coded pk/sk), in favor of using established interfaces that abstract the handling of particular queries. Here's an example:
https://github.com/InspiringApps/JCC-CompactConnect/blob/ff8c7c05c240d531f69c4502268337f607df5ff1/backend/compact-connect/lambdas/python/data-events/tests/function/test_encumbrance_events.py#L247

**Review by jusdino — CHANGES_REQUESTED** [review]
Looks good generally. You ran across a couple of items we've been trying to tidy up in the codebase that hopefully you can tidy up for us here:

**Comment by jusdino** [comment]
Also, now that we've converted to the git-tag workflow, you'll need to rebase your branch off from CSG's `main` and update this PR to be against `main`.

File: backend/compact-connect/lambdas/python/common/cc_common/data_model/data_client.py [hunk 3]
```diff
@@ -2905,7 +2905,7 @@ def encumber_home_jurisdiction_license_privileges(
         compact: str,
         provider_id: str,
         jurisdiction: str,
-        adverse_action_id: str,
+        adverse_action_id: UUID,
```
Comment (jusdino) [inline backend/compact-connect/lambdas/python/common/cc_common/data_model/data_client.py]: You'll also need to update the data events handler, where this gets called, to ensure that actually does pass in a UUID.

File: backend/compact-connect/lambdas/python/data-events/tests/function/test_encumbrance_events.py [hunk 4]
```diff
@@ -3033,6 +3033,239 @@ def test_license_encumbrance_lifting_notification_listener_skips_notifications_w
...
+        self.assertEqual(1, len(matching_updates))
+        first_update_record = matching_updates[0]
+        self.assertEqual(UpdateCategory.LIFTING_ENCUMBRANCE, first_update_record.updateType)
+        self.assertEqual({'encumberedStatus': 'unencumbered'}, first_update_record.updatedValues)
+
+        # Execute the handler SECOND TIME (simulating re-run of same event)
+        license_encumbrance_lifted_listener(event, self.mock_context)
+
+        # Verify STILL only one update record exists (no duplicate created)
+        # The function should return early because no LICENSE_ENCUMBERED privileges remain
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/data-events/tests/function/test_encumbrance_events.py]: A bit confused by this comment- can you clarify?
